src/NCalc/Parser/LogicalExpressionParser.cs

```diff
@@ -0,0 +1,254 @@
+#nullable disable
+
+using System;
+using System.Globalization;
+using NCalc.Domain;
+using NCalc.Exceptions;
+using Parlot.Fluent;
+using static Parlot.Fluent.Parsers;
+using Identifier = NCalc.Domain.Identifier;
+
+namespace NCalc.Parser;
+
+public static class LogicalExpressionParser
+{
+    private static readonly Parser<LogicalExpression> Parser;
+
+    private static readonly ValueExpression True = new(true);
+    private static readonly ValueExpression False = new(false);
+
+    static LogicalExpressionParser()
+    {
+        /*
+         * Grammar:
+         * expression     => ternary ( ( "-" | "+" ) ternary )* ;
+         * ternary        => equality ( "?" equality ":" equality)?
+         * equality       => relational ( ( "=" | "!=" | ... ) relational )* ;
+         * relational     => multiplicative ( ( ">=" | ">" | ... ) multiplicative )* ;
+         * multiplicative => unary ( ( "/" | "*" ) unary )* ;
+         * unary          => ( "-" | "not" ) unary
+         *                 | primary ;
+         *
+         * primary        => NUMBER
+         *                  | STRING
+         *                  | "true"
+         *                  | "false"
+         *                  | "[" anything "]"
+         *                  | function
+         *                  | "(" expression ")" ;
+         *
+         * function       => Identifier "(" arguments ")"
+         * arguments      => expression ( "," expression )*
+         */
+        // The Deferred helper creates a parser that can be referenced by others before it is defined
+        var expression = Deferred<LogicalExpression>();
+
+        var intParser = Terms.Integer().Then<LogicalExpression>(static d => new ValueExpression(d));
+
+        var decimalPointParser = Terms.Integer(NumberOptions.AllowSign)
+            .AndSkip(Terms.Char('.'))
+            .And(Terms.Integer())
+            .Then<LogicalExpression>((context, x) =>
+                new ValueExpression(((LogicalExpressionParserContext)context).UseDecimalsAsDefault
+                    ? Convert.ToDecimal(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture)
+                    : Convert.ToDouble(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture)));
+        
+        var trailingDecimalPointParser = Terms.Integer(NumberOptions.AllowSign)
+            .AndSkip(Terms.Char('.'))
+            .Then<LogicalExpression>((context, x) =>
+                new ValueExpression(((LogicalExpressionParserContext)context).UseDecimalsAsDefault
+                    ? Convert.ToDecimal(x, CultureInfo.InvariantCulture)
+                    : Convert.ToDouble(x, CultureInfo.InvariantCulture)));
```

Comment:
I think it can be combined into one with `ZeroOrOne`:

```
var decimalPointParser = Terms.Integer(NumberOptions.AllowSign)
    .AndSkip(Terms.Char('.'))
    .And(ZeroOrOne(Terms.Integer()))
    .Then<LogicalExpression>((context, x) =>
        new ValueExpression(((LogicalExpressionParserContext)context).UseDecimalsAsDefault
            ? Convert.ToDecimal(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture)
            : Convert.ToDouble(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture)));
```

Reply:
We should also handle scientific notification here, but it seems quite complicated

Reply:
Seems working fine (13 failing tests)

```
  var intParser = Terms.Integer(NumberOptions.AllowSign)
      .AndSkip(ZeroOrOne(Terms.Text("e", true)))
      .And(ZeroOrOne(Terms.Integer(NumberOptions.AllowSign)))
      .Then<LogicalExpression>((context, d) =>
      {
          if (d.Item2 != 0)
          {
              bool useDecimalAsDefault = ((LogicalExpressionParserContext)context).UseDecimalsAsDefault;

              if (useDecimalAsDefault)
              {
                  decimal val = Convert.ToDecimal(d.Item1 + "e" + d.Item2);
                  return new ValueExpression(val);
              }
              else
              {
                  double val = Convert.ToDouble(d.Item1 + "e" + d.Item2);
                  return new ValueExpression(val);
              }
          }

          return new ValueExpression(d.Item1);
      });

  var decimalPointParser = ZeroOrOne(Terms.Integer(NumberOptions.AllowSign))
      .AndSkip(Terms.Char('.'))
      .And(ZeroOrOne(Terms.Integer()))
      .AndSkip(ZeroOrOne(Terms.Text("e", true)))
      .And(ZeroOrOne(Terms.Integer(NumberOptions.AllowSign)))
      .Then<LogicalExpression>((context, x) =>
      {
          bool useDecimalAsDefault = ((LogicalExpressionParserContext)context).UseDecimalsAsDefault;

          if (useDecimalAsDefault)
          {
              decimal decimalValue = 0;
              if (x.Item3 != 0)
                  decimalValue = Convert.ToDecimal(x.Item1 + "." + x.Item2 + "e" + x.Item3, CultureInfo.InvariantCulture);
              else
                  decimalValue = Convert.ToDecimal(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture);

              return new ValueExpression(decimalValue);
          }
          else
          {
              double doubleValue = 0;
              if (x.Item3 != 0)
                  doubleValue = Convert.ToDouble(x.Item1 + "." + x.Item2 + "e" + x.Item3, CultureInfo.InvariantCulture);
              else
                  doubleValue = Convert.ToDouble(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture);

              return new ValueExpression(doubleValue);
          }
      });
```

Reply:
12 failing tests:

  ```
var intParser = Terms.Integer().Then<LogicalExpression>(static d => new ValueExpression(d));

  var intExponentParser = Terms.Integer(NumberOptions.AllowSign)
      .And(Terms.Text("e", true))
      .And(Terms.Integer(NumberOptions.AllowSign))
      .Then<LogicalExpression>((context, d) =>
      {
          if (!string.IsNullOrWhiteSpace(d.Item2))
          {
              bool useDecimalAsDefault = ((LogicalExpressionParserContext)context).UseDecimalsAsDefault;

              if (useDecimalAsDefault)
              {
                  decimal val = Convert.ToDecimal(d.Item1 + d.Item2 + d.Item3);
                  return new ValueExpression(val);
              }
              else
              {
                  double val = Convert.ToDouble(d.Item1 + d.Item2 + d.Item3);
                  return new ValueExpression(val);
              }
          }

          return new ValueExpression(d.Item1);
      });

  var decimalPointParser = ZeroOrOne(Terms.Integer(NumberOptions.AllowSign))
      .AndSkip(Terms.Char('.'))
      .And(ZeroOrOne(Terms.Integer()))
      .AndSkip(ZeroOrOne(Terms.Text("e", true)))
      .And(ZeroOrOne(Terms.Integer(NumberOptions.AllowSign)))
      .Then<LogicalExpression>((context, x) =>
      {
          bool useDecimalAsDefault = ((LogicalExpressionParserContext)context).UseDecimalsAsDefault;

          if (useDecimalAsDefault)
          {
              decimal decimalValue = 0;
              if (x.Item3 != 0)
                  decimalValue = Convert.ToDecimal(x.Item1 + "." + x.Item2 + "e" + x.Item3, CultureInfo.InvariantCulture);
              else
                  decimalValue = Convert.ToDecimal(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture);

              return new ValueExpression(decimalValue);
          }
          else
          {
              double doubleValue = 0;
              if (x.Item3 != 0)
                  doubleValue = Convert.ToDouble(x.Item1 + "." + x.Item2 + "e" + x.Item3, CultureInfo.InvariantCulture);
              else
                  doubleValue = Convert.ToDouble(x.Item1 + "." + x.Item2, CultureInfo.InvariantCulture);

              return new ValueExpression(doubleValue);
          }
      });

  var number = OneOf(
      decimalPointParser,
      intExponentParser,
      intParser
  );
```